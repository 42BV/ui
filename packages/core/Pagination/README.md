# @42.nl/ui-core-pagination

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-pagination

// with yarn
yarn add @42.nl/ui-core-pagination
```
