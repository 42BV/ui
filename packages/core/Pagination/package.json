{
  "name": "@42.nl/ui-core-pagination",
  "version": "0.0.1",
  "description": "",
  "main": "lib/Pagination.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "yarn clean && tsc && yarn copy:scss",
    "copy:scss": "rsync -a src/*.scss lib/scss/",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "mad-spring-connect": "3.1.1",
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "reactstrap": "8.0.1",
    "react": "16.8.6",
    "@storybook/react": "next"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
