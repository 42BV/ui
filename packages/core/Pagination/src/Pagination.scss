@import '~@42.nl/ui-core-styling/scss/variables';
@import '~@42.nl/ui-core-styling/scss/bootstrap';

.pagination {
  .page-item {
    button.disabled.page-link {
      border-radius: 0;
      background-color: $white;
      color: $body-color;
    }

    &.active {
      .page-link {
        overflow: hidden;
      }
    }
  }
}
