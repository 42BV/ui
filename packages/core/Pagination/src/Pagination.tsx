import React from 'react';
import { Page } from 'mad-spring-connect';

import {
  Pagination as RPagination,
  PaginationItem,
  PaginationLink
} from 'reactstrap';

interface Props<T> {
  /**
   * Represents Spring's page abstraction.
   */
  page: Page<T>;

  /**
   * Called when navigation to a certain page number.
   */
  onChange: (pageNumber: number) => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

/**
 * The Pagination component is an enhanced Bootstrap pagination component. It supports working with `Page`s and shows a fixed
 * layout for the number of pages, indicating that multiple pages exist for the min and max ranges (e.g. 1 ... 4 5 6 ... 10).
 */
export default function Pagination<T>({ page, onChange, className }: Props<T>) {
  const { first, last, totalPages } = page;

  // Don't bother to render if there is nothing to paginate.
  if (first && last) {
    return null;
  }

  const current = page.number;
  const content = pagesFor(current, totalPages);

  return (
    <RPagination className={className}>
      {content.map((item, index) => (
        <PaginationItem active={item === current} key={index}>
          {item !== '...' ? (
            <PaginationLink onClick={() => onChange(item)}>
              {item}
            </PaginationLink>
          ) : (
            <PaginationLink className="disabled" disabled={true}>
              ...
            </PaginationLink>
          )}
        </PaginationItem>
      ))}
    </RPagination>
  );
}

type Dots = '...';

// Calculates which pages surround the current page.
export function pagesFor(
  currentPage: number,
  totalPages: number
): (number | Dots)[] {
  const content = [];

  if (currentPage > 1) {
    content.push(1);
  }

  const prev = currentPage - 1;
  if (prev > 1) {
    content.push(prev);
  }

  content.push(currentPage);

  const next = currentPage + 1;
  if (next <= totalPages) {
    content.push(next);
  }

  if (next + 1 <= totalPages) {
    content.push(totalPages);
  }

  if (content[0] === 1 && content[1] !== 2) {
    // @ts-ignore
    content.splice(1, 0, '...');
  }

  const lastIndex = content.length - 1;
  if (
    content[lastIndex] === totalPages &&
    content[lastIndex - 1] !== totalPages - 1
  ) {
    // @ts-ignore
    content.splice(lastIndex, 0, '...');
  }

  return content;
}
