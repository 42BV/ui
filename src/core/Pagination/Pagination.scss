.pagination {
  margin-bottom: 0;

  .page-item {
    button.disabled.page-link {
      border-radius: 0;
      background-color: $white;
      color: $body-color;
    }

    &.active {
      .page-link {
        overflow: hidden;
      }
    }
  }
}
