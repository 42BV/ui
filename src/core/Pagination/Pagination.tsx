import React from 'react';
import { Page } from '@42.nl/spring-connect';

import {
  Pagination as RPagination,
  PaginationItem,
  PaginationLink
} from 'reactstrap';

import { Icon } from '../Icon';
import { range } from 'lodash';

type Props<T> = {
  /**
   * Represents Spring's page abstraction.
   */
  page: Page<T>;

  /**
   * Called when navigation to a certain page number.
   */
  onChange: (pageNumber: number) => void;

  /**
   * Whether or not to show the previous and next buttons.
   *
   * Defaults to `true`
   */
  showPreviousAndNextButtons?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optionally customized text to use within the component.
   */
  text?: Text;
};

/**
 * The Pagination component is an enhanced Bootstrap pagination component. It supports working with `Page`s and shows a fixed
 * layout for the number of pages, indicating that multiple pages exist for the min and max ranges (e.g. 1 ... 4 5 6 ... 10).
 */
export default function Pagination<T>({
  page,
  onChange,
  className,
  showPreviousAndNextButtons = true
}: Props<T>) {
  const { first, last, totalPages } = page;

  // Don't bother to render if there is nothing to paginate.
  if (first && last) {
    return null;
  }

  const current = page.number;
  const content = pagesFor(current, totalPages);

  return (
    <RPagination className={className}>
      {showPreviousAndNextButtons ? (
        <PaginationItem disabled={first}>
          <PaginationLink onClick={() => onChange(current - 1)}>
            <Icon icon="arrow_back" size={14} />
          </PaginationLink>
        </PaginationItem>
      ) : null}
      {content.map((item, index) => (
        <PaginationItem active={item === current} key={index}>
          {item !== '...' ? (
            <PaginationLink onClick={() => onChange(item)}>
              {item}
            </PaginationLink>
          ) : (
            <PaginationLink className="disabled" disabled={true}>
              ...
            </PaginationLink>
          )}
        </PaginationItem>
      ))}
      {showPreviousAndNextButtons ? (
        <PaginationItem disabled={last}>
          <PaginationLink onClick={() => onChange(current + 1)}>
            <Icon icon="arrow_forward" size={14} />
          </PaginationLink>
        </PaginationItem>
      ) : null}
    </RPagination>
  );
}

type Ellipsis = '...';

// Calculates which pages surround the current page.
export function pagesFor(
  currentPage: number,
  totalPages: number
): (number | Ellipsis)[] {
  const ellipsis: Ellipsis = '...';

  if (totalPages < 8) {
    return range(1, totalPages + 1);
  }

  if (currentPage < 5) {
    // {1} 2 3 ... 29 30 31
    // till
    // 1 2 3 {4} 5 ... 31

    const tillEllipsis = range(1, Math.max(4, currentPage + 2));
    const fromEllipsis = range(totalPages - (5 - tillEllipsis.length), totalPages + 1)

    return [ ...tillEllipsis, ellipsis, ...fromEllipsis ];
  }

  if (currentPage > totalPages - 4) {
    // 1 ... 27 {28} 29 30 31
    // till
    // 1 2 3 ... 29 30 {31}

    const fromEllipsis = range(Math.min(totalPages - 2, currentPage - 1), totalPages + 1);
    const tillEllipsis = range(1, 7 - fromEllipsis.length);

    return [ ...tillEllipsis, ellipsis, ...fromEllipsis ];
  }

  return [1, ellipsis, ...range(currentPage - 1, currentPage + 2), ellipsis, totalPages];
}
