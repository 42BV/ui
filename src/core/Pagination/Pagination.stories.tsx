import React, { useState } from 'react';
import { range } from 'lodash';

import { pageOf } from '../../utilities/page/page';

import { Pagination } from './Pagination';
import { Alert } from 'reactstrap';

export default {
  title: 'core/Pagination',

  decorators: [
    (Story) => (
      <>
        <Alert color="warning" className="mb-4">
          <p>
            To be able to use Pagination, you have to add @42.nl/spring-connect
            and lodash to your dependencies:
          </p>
          <code>npm install --save @42.nl/spring-connect lodash</code>
        </Alert>
        <Story />
      </>
    )
  ],

  parameters: {
    component: Pagination
  }
};

const DefaultStory = () => {
  const [pageNumber, setPageNumber] = useState(5);

  const page = pageOf(range(1, 100), pageNumber, 10);

  return (
    <div className="d-flex justify-content-center">
      <Pagination page={page} onChange={setPageNumber} />
    </div>
  );
};

export const Default = {
  render: DefaultStory,
  name: 'default'
};

const WithoutPreviousAndNextStory = () => {
  const [pageNumber, setPageNumber] = useState(5);

  const page = pageOf(range(1, 100), pageNumber, 10);

  return (
    <div className="d-flex justify-content-center">
      <Pagination
        page={page}
        onChange={setPageNumber}
        showPreviousAndNextButtons={false}
      />
    </div>
  );
};

export const WithoutPreviousAndNext = {
  render: WithoutPreviousAndNextStory,
  name: 'without previous and next'
};

const WithChangeablePageSizeStory = () => {
  const [pageNumber, setPageNumber] = useState(5);
  const [pageSize, setPageSize] = useState(10);

  const page = pageOf(range(1, 200), pageNumber, pageSize);

  return (
    <div className="d-flex justify-content-center">
      <Pagination
        page={page}
        onChange={setPageNumber}
        onPageSizeChange={(size) => {
          setPageSize(size);
          setPageNumber(1);
        }}
      />
    </div>
  );
};

export const WithChangeablePageSize = {
  render: WithChangeablePageSizeStory,
  name: 'with changeable page size'
};

const WithAllowedPageSizesStory = () => {
  const [pageNumber, setPageNumber] = useState(5);
  const [pageSize, setPageSize] = useState(8);

  const page = pageOf(range(1, 200), pageNumber, pageSize);

  return (
    <div className="d-flex justify-content-center">
      <Pagination
        page={page}
        onChange={setPageNumber}
        onPageSizeChange={(size) => {
          setPageSize(size);
          setPageNumber(1);
        }}
        allowedPageSizes={[2, 4, 8, 16, 32]}
      />
    </div>
  );
};

export const WithAllowedPageSizes = {
  render: WithAllowedPageSizesStory,
  name: 'with allowed page sizes'
};
