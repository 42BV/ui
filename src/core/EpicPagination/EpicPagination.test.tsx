import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Page } from '@42.nl/spring-connect';

import Pagination, { pagesFor } from './EpicPagination';

describe('Component: Pagination', () => {
  function setup({
    number,
    totalPages
  }: {
    number: number;
    totalPages: number;
  }) {
    const page: Page<any> = {
      content: [],
      last: number === totalPages,
      totalElements: 100,
      totalPages,
      size: 10,
      number,
      first: number === 1,
      numberOfElements: 10
    };

    const onChangeSpy = jest.fn();

    const props = {
      page,
      onChange: onChangeSpy
    };

    const { container } = render(<Pagination {...props} />);
    
    return { container, onChangeSpy };
  }

  describe('ui', () => {
    test('empty', () => {
      const { container } = setup({ number: 1, totalPages: 1 });

      expect(container.firstChild).toBeNull();
    });

    test('middle', () => {
      const { container } = setup({ number: 5, totalPages: 10});
      expect(container).toMatchSnapshot();
      expect(screen.queryByText('1')).not.toBeInTheDocument();
      expect(screen.queryByText('3')).toBeInTheDocument();
      expect(screen.getByText('5').parentNode).toHaveClass('active');
      expect(screen.queryByText('7')).toBeInTheDocument();
      expect(screen.queryByText('10')).not.toBeInTheDocument();
      expect(screen.getByText('first_page').parentNode?.parentNode).not.toHaveClass('disabled');
      expect(screen.getByText('last_page').parentNode?.parentNode).not.toHaveClass('disabled');
      expect(screen.getByText('chevron_left').parentNode?.parentNode).not.toHaveClass('disabled');
      expect(screen.getByText('chevron_right').parentNode?.parentNode).not.toHaveClass('disabled');
    });

    test('first page', () => {
      setup({ number: 1, totalPages: 10});
      expect(screen.queryByText('1')).toBeInTheDocument();
      expect(screen.getByText('1').parentNode).toHaveClass('active');
      expect(screen.queryByText('3')).toBeInTheDocument();
      expect(screen.queryByText('4')).not.toBeInTheDocument();
      expect(screen.getByText('first_page').parentNode?.parentNode).toHaveClass('disabled');
      expect(screen.getByText('chevron_left').parentNode?.parentNode).toHaveClass('disabled');
    });

    test('last page', () => {
      setup({ number: 10, totalPages: 10});
      expect(screen.queryByText('10')).toBeInTheDocument();
      expect(screen.getByText('10').parentNode).toHaveClass('active');
      expect(screen.queryByText('8')).toBeInTheDocument();
      expect(screen.queryByText('7')).not.toBeInTheDocument();
      expect(screen.getByText('last_page').parentNode?.parentNode).toHaveClass('disabled');
      expect(screen.getByText('chevron_right').parentNode?.parentNode).toHaveClass('disabled');
    });
  });

  describe('large numbers', () => {
    it('should decrease font size when page number over 1000', () => {
      setup({number: 1002, totalPages: 100000});
      expect(screen.getByText('1002')).toHaveClass('scale-to-page-size-1000');
    });

    it('should decrease font size even more when page number over 10000', () => {
      setup({number: 10002, totalPages: 100000});
      expect(screen.getByText('10002')).toHaveClass('scale-to-page-size-10000');
    });
  });

  describe('events', () => {
    it('should call onChange when pagination links are clicked', () => {
      const { onChangeSpy } = setup({ number: 5, totalPages: 10 });

      fireEvent.click(screen.getByText('first_page'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(1);

      fireEvent.click(screen.getByText('4'));

      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      expect(onChangeSpy).toHaveBeenCalledWith(4);

      fireEvent.click(screen.getByText('5'));

      expect(onChangeSpy).toHaveBeenCalledTimes(3);
      expect(onChangeSpy).toHaveBeenCalledWith(5);

      fireEvent.click(screen.getByText('6'));

      expect(onChangeSpy).toHaveBeenCalledTimes(4);
      expect(onChangeSpy).toHaveBeenCalledWith(6);

      fireEvent.click(screen.getByText('last_page'));

      expect(onChangeSpy).toHaveBeenCalledTimes(5);
      expect(onChangeSpy).toHaveBeenCalledWith(10);
    });

    describe('previous and next buttons', () => {
      it('should have previous and next buttons which move the page forward or backwards', () => {
        const { onChangeSpy } = setup({ number: 2, totalPages: 3 });

        fireEvent.click(screen.getByText('chevron_left'));

        expect(onChangeSpy).toHaveBeenCalledTimes(1);
        expect(onChangeSpy).toHaveBeenCalledWith(1);

        fireEvent.click(screen.getByText('chevron_right'));

        expect(onChangeSpy).toHaveBeenCalledTimes(2);
        expect(onChangeSpy).toHaveBeenCalledWith(3);
      });
    });
  });
});

describe('pages', () => {
  test('exact middle', () => {
    expect(pagesFor(5, 10)).toEqual([3, 4, 5, 6, 7]);
  });
  
  test('almost middle', () => {
    expect(pagesFor(4, 10)).toEqual([2, 3, 4, 5, 6]);
    expect(pagesFor(6, 10)).toEqual([4, 5, 6, 7, 8]);
  });
  
  test('absolute left', () => {
    expect(pagesFor(1, 10)).toEqual(['...', '...', 1, 2, 3]);
  });
  
  test('almost left', () => {
    expect(pagesFor(2, 10)).toEqual(["...", 1, 2, 3, 4]);
    expect(pagesFor(3, 10)).toEqual([1, 2, 3, 4, 5]);
    expect(pagesFor(4, 10)).toEqual([2, 3, 4, 5, 6]);
  });
  
  test('almost right', () => {
    expect(pagesFor(7, 10)).toEqual([5, 6, 7, 8, 9]);
    expect(pagesFor(8, 10)).toEqual([6, 7, 8, 9, 10]);
    expect(pagesFor(9, 10)).toEqual([7, 8, 9, 10,'...']);
  });
  
  test('absolute right', () => {
    expect(pagesFor(10, 10)).toEqual([8, 9, 10, '...', '...']);
  });
  
  test('shortage', () => {
    expect(pagesFor(1, 3)).toEqual(['...', '...', 1, 2, 3]);
    expect(pagesFor(3, 3)).toEqual([1, 2, 3, '...', '...']);
  });
 }); 
