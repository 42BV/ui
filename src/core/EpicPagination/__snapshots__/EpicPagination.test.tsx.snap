// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: EpicPagination ui middle 1`] = `
<div>
  <div
    class="d-flex justify-content-center flex-wrap flex-sm-nowrap"
  >
    <nav
      aria-label="pagination"
      class="epic-pagination"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            <i
              class="icon center-icon material-icons"
            >
              first_page
            </i>
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            <i
              class="icon center-icon material-icons"
            >
              chevron_left
            </i>
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            3
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            4
          </button>
        </li>
        <li
          class="page-item active"
        >
          <button
            class="page-link"
          >
            5
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            6
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            7
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            <i
              class="icon center-icon material-icons"
            >
              chevron_right
            </i>
          </button>
        </li>
        <li
          class="page-item"
        >
          <button
            class="page-link"
          >
            <i
              class="icon center-icon material-icons"
            >
              last_page
            </i>
          </button>
        </li>
        <li
          class="page-item disabled"
        >
          <div
            class="total-elements"
          >
            100
             records found
          </div>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
